r-reproducibility.Rmd

```diff
@@ -0,0 +1,1211 @@
+and easier for you to use later on.
+
+We also use the `here()` function when we import a dataset or save a dataset.
+So, let's load in the NYC Dog License dataset. First, save the csv in the
+`data/` folder. Then create a new R file:
+
+```r
+usethis::use_r("load-data")
+```
+
+And write these lines in the file:
+
+```{r load-data-example, eval=FALSE}
+source(here::here("R/package-loading.R"))
+dog_license <- read_csv(here::here("data/nyc-dog.....csv"))
+# TODO complete this file path.
+head(dog_license)
+```
+
+That is how we will load data in from now on.
+
+Here are a few other tips for keeping your code reusable:
+
+- Aggressively try keep your code as simple and concise as you can. The less code
+you have to write, debug, and maintain, the better. If you repeat some code,
+stop and think how to not repeat it. You will save time in the end by taking
+time to think and plan.
+- When encountering a difficult problem, think about how to solve it, then try
+to find R packages or functions that do your problem for you. You may hear some
+poeple say "oh, don't bother with R packages, do everything in base R"... don't 
+listen to them. Do you build a computer from scratch everytime you want to do
+any type of work? Or a car when you want to drive somewhere? No, you don't. Make
+use of other people's hard work to make tools that simplify your life.
+- Split up your analyses steps into individual files (e.g. "model" file, "plot"
+file). Then `source` those files as needed or save the output in `data/` to use 
+it in other files.
+- Try not to have R scripts be too long (e.g. more than 500 lines of code). Keep
+script sizes as small as necessary and as specific as possible (have a single
+purpose). A script should have an end goal.
+
+## Readability
+
+There are two reasons we write code: to instruct the computer to do something and
+to record the steps we took to get a particular result for us or others to
+understand. For computers, *how* or *what* you write doesn't matter, as long as
+the code is correct. Computers don't need to *understand* the code. But humans
+do need to understand it. We need clear language and explicit meaning in order
+to understand what is going on. Humans write code, humans read code, and humans
+must maintain them and fix any errors. So, what you write and how you write it
+is extremely important.
+
+Like natural human languages, R has a relaxed approached to how R code is written.
+This has some nice advantages, but also some major disadvantages, notably that
+writing styles can be quite different across the world or even within one's own
+code. So, it's important to stick to some guidelines, for instance, as laid out
+by the tidyverse [style guide][r-style-guide].
+
+Some tips to use when working in R:
+
+- Think of writing code as if you are writing in a natural language like English.
+- Imagine that other people will be reading your code.
+- Keep the code clear, simple, and *readable*.
```

Comment:
Again, nobody sets out to write *unreadable* code, but if they don't have concrete examples or guidelines, they won't know what "readable" is.